### PR TITLE
bitbox: simplify writeout

### PR DESCRIPTION
**Pull request by gabriele-0201**
I thought that `IoDmux` was a bit over-complicated because we don't really need to worry about the order of writes during writeout; we just need them to occur.

An alternative to this pr is to modify `IoDmux` and use the same approach used in `LnWriteOut` and `BbnWriteOut` for `MetaSwap`

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/270?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#270** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/270?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#269** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/269?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/beatree/writeout.rs [hunk 1]
```diff
@@ -74,34 +71,25 @@ fn do_run(mut cx: Cx, mut io: IoDmux) {
...
-        // At this point, the BBN and LN files are fully synced. We can now update the meta file.
-        if let Some(ref mut meta_swap) = &mut cx.meta_swap {
-            if meta_swap.run(&mut io) {
-                cx.meta_swap = None;
-                return;
-            }
-        }
-    }
+    cx.bbn_write_out.wait_fsync(&mut io);
+    cx.ln_write_out.wait_fsync(&mut io);
```
Comment (rphmeier) [inline nomt/src/beatree/writeout.rs:78]: seems better to dispatch the fsyncs at the same time and then handle the completions async.

for a follow-up.

**Review by rphmeier — CHANGES_REQUESTED** [review]
I'm a bit conflicted around this PR.

On one hand, it does simplify things and looks like it is more aggressive with pushing onto io_uring. On the other, it looks like actually handling any kind of io error is going to be nearly impossible and would require either adding something like the inboxes back or to encapsulate the logic within `fn recv`. The inbox approach was better for that. Likewise, any attempt to handle completions _while_ write submission is ongoing would require another refactor of this code.

User data on io requests/completions is useful!

Even though it's simpler in the sense that there's less code, it's more conceptually complicated. 

The only refactor that was needed IMO is to move completion polling out of the `run` functions and have each helper gain a `handle_completion` function. And to have the main `run` loop attempt to take a single completion on each iteration, so long as each `run` indicated it wasn't currently trying to submit anything / the completion queue was full.

**Comment by pepyakin** [comment]
Agreed. Closing for now just to clean up the queue, but feel free to resurrect as soon as you come back @gabriele-0201.

**Pull request closed without merging**